File: ocean_stats.py
#!/usr/bin/env python3
import requests
import json
import os

def fetch_ocean_data():
    """
    Fetches and displays mining data for the OCEAN pool using the mempool.space API.
    """
    # API Endpoints for the OCEAN pool
    base_url = "https://mempool.space/api/v1/mining"
    slug = "ocean"

    endpoints = {
        "Pool Details": f"{base_url}/pool/{slug}",
        "Hashrate History": f"{base_url}/pool/{slug}/hashrate",
        "Recent Blocks": f"{base_url}/pool/{slug}/blocks"
    }

    print(f"--- Querying mempool.space for Pool: {slug.upper()} ---")

    for title, url in endpoints.items():
        try:
            response = requests.get(url)
            response.raise_for_status()
            data = response.json()

            print(f"\n[+] {title}:")
            # Print a formatted preview of the data
            if isinstance(data, list):
                # Print first 2 items if it's a list (e.g., blocks or hashrate history)
                print(json.dumps(data[:2], indent=4))
                print(f"... ({len(data)} total items returned)")
            else:
                print(json.dumps(data, indent=4))

        except requests.exceptions.RequestException as e:
            print(f"[-] Error fetching {title}: {e}")

def generate_ocean_config_env():
    """
    Sets local environment variables for the BIP-64MOD + GCC context
    using the official pool address found in the documentation.
    """
    os.environ["POOL_URL"] = "mine.ocean.xyz:3334"
    os.environ["POOL_API_SLUG"] = "ocean"

    print("\n--- Local Environment Configured ---")
    print(f"POOL_URL: {os.environ.get('POOL_URL')}")
    print(f"API_SLUG: {os.environ.get('POOL_API_SLUG')}")

if __name__ == "__main__":
    # Ensure dependencies are available
    try:
        import requests
    except ImportError:
        print("Error: 'requests' library not found. Install it with: pip install requests")
        exit(1)

    # 1. Configure the local environment
    generate_ocean_config_env()

    # 2. Fetch live data from the API
    fetch_ocean_data()

File: src/bin/ocean_stats.rs
use reqwest;
use serde::Deserialize;
use anyhow::{Result, Context};
use serde_json::Value;
use std::env;

// Structs for partial deserialization if needed, otherwise use Value for flexibility
#[derive(Debug, Deserialize)]
struct PoolDetails {
    // Assuming some fields, but using Value for the rest
    #[serde(flatten)]
    data: Value,
}

#[derive(Debug, Deserialize)]
struct Block {
    height: u64,
    // We'll just take the height for display, rest can be Value if not used
    #[serde(flatten)]
    data: Value,
}

async fn fetch_ocean_data_rust() -> Result<()> {
    let base_url = "https://mempool.space/api/v1/mining";
    let slug = "ocean";

    let endpoints = [
        ("Pool Details", format!("{}/pool/{}", base_url, slug)),
        ("Hashrate History", format!("{}/pool/{}/hashrate", base_url, slug)),
        ("Recent Blocks", format!("{}/pool/{}/blocks", base_url, slug)),
    ];

    println!("--- Querying mempool.space for Pool: {} ---", slug.to_uppercase());

    for (title, url) in endpoints.into_iter() {
        match reqwest::get(&url).await {
            Ok(response) => {
                response.raise_for_status().context(format!("HTTP error for {}", title))?;
                let data: Value = response.json().await?;

                println!("
[+] {}:", title);
                if data.is_array() {
                    // Print first 2 items if it's a list
                    if let Some(arr) = data.as_array() {
                        let preview: Vec<&Value> = arr.iter().take(2).collect();
                        println!("{}", serde_json::to_string_pretty(&preview)?);
                        println!("... ({} total items returned)", arr.len());
                    }
                } else {
                    println!("{}", serde_json::to_string_pretty(&data)?);
                }
            },
            Err(e) => {
                eprintln!("[-] Error fetching {}: {}", title, e);
            }
        }
    }

    Ok(())
}

fn generate_ocean_config_env_rust() {
    env::set_var("POOL_URL", "mine.ocean.xyz:3334");
    env::set_var("POOL_API_SLUG", "ocean");

    println!("
--- Local Environment Configured ---");
    println!("POOL_URL: {}", env::var("POOL_URL").unwrap_or_else(|_| "N/A".to_string()));
    println!("API_SLUG: {}", env::var("POOL_API_SLUG").unwrap_or_else(|_| "N/A".to_string()));
}

#[tokio::main]
async fn main() -> Result<()> {
    // 1. Configure the local environment
    generate_ocean_config_env_rust();

    // 2. Fetch live data from the API
    fetch_ocean_data_rust().await
}
